dice_ml/explainer_interfaces/dice_KD.py

```diff
@@ -0,0 +1,236 @@
+
+        :param algorithm: Counterfactual generation algorithm. Either "DiverseCF" or "RandomInitCF".
+        :param features_to_vary: Either a string "all" or a list of feature names to vary.
+        param permitted_range: Dictionary with continuous feature names as keys and permitted min-max range in list as values. Defaults to the range inferred from training data. If None, uses the parameters initialized in data_interface.
+        :param yloss_type: Metric for y-loss of the optimization function. Takes "l2_loss" or "log_loss" or "hinge_loss".
+        :param diversity_loss_type: Metric for diversity loss of the optimization function. Takes "avg_dist" or "dpp_style:inverse_dist".
+        :param feature_weights: Either "inverse_mad" or a dictionary with feature names as keys and corresponding weights as values. Default option is "inverse_mad" where the weight for a continuous feature is the inverse of the Median Absolute Devidation (MAD) of the feature's values in the training set; the weight for a categorical feature is equal to 1 by default.
+        :param optimizer: Tensorflow optimization algorithm. Currently tested only with "tensorflow:adam".
+
+        :param learning_rate: Learning rate for optimizer.
+        :param min_iter: Min iterations to run gradient descent for.
+        :param max_iter: Max iterations to run gradient descent for.
+        :param project_iter: Project the gradients at an interval of these many iterations.
+        :param loss_diff_thres: Minimum difference between successive loss values to check convergence.
+        :param loss_converge_maxiter: Maximum number of iterations for loss_diff_thres to hold to declare convergence. Defaults to 1, but we assigned a more conservative value of 2 in the paper.
+        :param verbose: Print intermediate loss value.
+        :param init_near_query_instance: Boolean to indicate if counterfactuals are to be initialized near query_instance.
+        :param tie_random: Used in rounding off CFs and intermediate projection.
+        :param stopping_threshold: Minimum threshold for counterfactuals target class probability.
+        :param posthoc_sparsity_param: Parameter for the post-hoc operation on continuous features to enhance sparsity.
+        :param posthoc_sparsity_algorithm: Perform either linear or binary search. Takes "linear" or "binary". Prefer binary search when a feature range is large (for instance, income varying from 10k to 1000k) and only if the features share a monotonic relationship with predicted outcome in the model.
+
+        :return: A CounterfactualExamples object to store and visualize the resulting counterfactual explanations (see diverse_counterfactuals.py).
+
+        """
+
+        # check feature MAD validity and throw warnings
+        if feature_weights == "inverse_mad":
+            self.data_interface.get_valid_mads(display_warnings=True, return_mads=False)
+
+        # check permitted range for continuous features
```

Comment:
code for checking permitted range should move to a base class explainerbase, since it can apply to all methods and we want to avoid repeating code. do you want to take a go at it?